source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -128,30 +159,91 @@ Network::FilterStatus UdpProxyFilter::ClusterInfo::onData(Network::UdpRecvData&
   }
 
   active_session->write(*data.buffer_);
-
   return Network::FilterStatus::StopIteration;
 }
 
-UdpProxyFilter::ActiveSession*
-UdpProxyFilter::ClusterInfo::createSession(Network::UdpRecvData::LocalPeerAddresses&& addresses,
-                                           const Upstream::HostConstSharedPtr& host) {
-  auto new_session = std::make_unique<ActiveSession>(*this, std::move(addresses), host);
-  auto new_session_ptr = new_session.get();
-  sessions_.emplace(std::move(new_session));
-  host_to_sessions_[host.get()].emplace(new_session_ptr);
-  return new_session_ptr;
+UdpProxyFilter::ActiveSession* UdpProxyFilter::StickySessionClusterInfo::getSession(
+    const Network::UdpRecvData::LocalPeerAddresses& addresses,
+    const Upstream::HostConstSharedPtr&) const {
+  const auto active_session_it = sessions_.find(addresses);
+  if (active_session_it != sessions_.end()) {
+    return active_session_it->get();
+  }
+  return nullptr;
 }
 
-void UdpProxyFilter::ClusterInfo::removeSession(const ActiveSession* session) {
-  // First remove from the host to sessions map.
-  ASSERT(host_to_sessions_[&session->host()].count(session) == 1);
-  auto host_sessions_it = host_to_sessions_.find(&session->host());
-  host_sessions_it->second.erase(session);
-  if (host_sessions_it->second.empty()) {
-    host_to_sessions_.erase(host_sessions_it);
+void UdpProxyFilter::StickySessionClusterInfo::storeSession(ActiveSessionPtr session) {
+  sessions_.emplace(std::move(session));
+}
+
+void UdpProxyFilter::StickySessionClusterInfo::removeSessionFromStorage(
+    const ActiveSession* session) {
+  ASSERT(sessions_.count(session) == 1);
+  sessions_.erase(session);
+}
+
+UdpProxyFilter::PerPacketLoadBalancingClusterInfo::PerPacketLoadBalancingClusterInfo(
+    UdpProxyFilter& filter, Upstream::ThreadLocalCluster& cluster)
+    : ClusterInfo(filter, cluster) {}
+
+UdpProxyFilter::PerPacketLoadBalancingClusterInfo::~PerPacketLoadBalancingClusterInfo() {
+  // Sanity check the session accounting. This is not as fast as a straight teardown, but this is
+  // not a performance critical path.
+  while (!sessions_.empty()) {
+    removeSession(sessions_.begin()->get());
+  }
+}
+
+Network::FilterStatus
+UdpProxyFilter::PerPacketLoadBalancingClusterInfo::onData(Network::UdpRecvData& data) {
+  UdpLoadBalancerContext context(filter_.config_->hashPolicy(), data.addresses_.peer_);
+  Upstream::HostConstSharedPtr host = cluster_.loadBalancer().chooseHost(&context);
+  if (host == nullptr) {
+    ENVOY_LOG(debug, "cannot find any valid host. failed to create a session.");
+    cluster_.info()->stats().upstream_cx_none_healthy_.inc();
+    return Network::FilterStatus::StopIteration;
+  }
+
+  ENVOY_LOG(debug, "selected {} host as upstream.", host->address()->asStringView());
+
+  auto active_session = getSession(data.addresses_, host);
```

Comment:
There is still quite a bit of duplicated code in here? I think very few lines should need to change between the two types? Can you try to collapse it further?

/wait

Reply:
done with previous commit